xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+          logger.log(Level.SEVERE, "Exception!" + e);
+          // TODO(chengyuanzhang): implement cleanup.
+        }
+      });
+
+  private final String listenerAddress;
+  private final ServerBuilder<?> delegateBuilder;
+  private ScheduledExecutorService timeService = SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE);
+  private final long retryDelayNano;
+  private final XdsClientPoolFactory xdsClientPoolFactory;
+  private final XdsServingStatusListener listener;
+  private final AtomicReference<FilterChainSelector> filterChainSelectorRef;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private final AtomicBoolean shutdown = new AtomicBoolean(false);
+  private boolean isServing;
+  private final CountDownLatch internalTerminationLatch = new CountDownLatch(1);
+  private final SettableFuture<IOException> initialStartFuture = SettableFuture.create();
+  private boolean initialStarted;
+  private ScheduledHandle restartTimer;
+  private ObjectPool<XdsClient> xdsClientPool;
+  private XdsClient xdsClient;
+  private DiscoveryState discoveryState;
+  private volatile Server delegate;
+
+  XdsServerWrapper(
+      String listenerAddress,
+      ServerBuilder<?> delegateBuilder,
+      long retryDelayNano,
+      XdsServingStatusListener listener,
+      AtomicReference<FilterChainSelector> filterChainSelectorRef,
+      XdsClientPoolFactory xdsClientPoolFactory) {
+    this.listenerAddress = checkNotNull(listenerAddress, "listenerAddress");
+    this.delegateBuilder = checkNotNull(delegateBuilder, "delegateBuilder");
+    this.retryDelayNano = retryDelayNano;
+    this.listener = checkNotNull(listener, "listener");
+    this.filterChainSelectorRef = checkNotNull(filterChainSelectorRef, "filterChainSelectorRef");
+    this.xdsClientPoolFactory = checkNotNull(xdsClientPoolFactory, "xdsClientPoolFactory");
+    this.delegate = delegateBuilder.build();
+  }
+
+  @Override
+  public Server start() throws IOException {
+    checkState(started.compareAndSet(false, true), "Already started");
+    syncContext.execute(new Runnable() {
+      @Override
+      public void run() {
+        internalStart();
+      }
+    });
+    syncContext.drain();
+    IOException exception;
+    try {
+      exception = initialStartFuture.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (exception != null) {
+      throw exception;
+    }
+    return this;
+  }
+
+  private void internalStart() {
+    try {
+      xdsClientPool = xdsClientPoolFactory.getOrCreate();
+    } catch (Exception e) {
+      listener.onNotServing(
+          Status.UNAVAILABLE.withDescription(
+              "Failed to initialize xDS").withCause(e).asException());
+      initialStartFuture.set(null);
+      return;
+    }
+    xdsClient = xdsClientPool.getObject();
+    // TODO(chengyuanzhang): add an API on XdsClient indicating if it is using v3, don't get
```

Comment:
Is the TODO still for chengyuanzhang ?

Reply:
Also when would they be different i.e. `isUseProtocolV3()` value from bootstrap vs XdsClient indicating if it's using v3. If not we should look into getting this from the bootstrap even before the xdsClient is started - why wait for xdsClient to start?

Reply:
ok